### ISSUE TITLE
IsolatedStorage table locking

### ISSUE THREAD

**Issue by pzentner** [microsoft/AL#4871]
Using microsoft/bcsandbox:us version tag 14.0.29537.31096. AL compiler version 3.0.106655.

I have started using IsolatedStorage to store certain information about the app. When I use  the following code:

```
        if not IsolatedStorage.Contains(KeyName) then
            exit('');

        if not IsolatedStorage.Get(KeyName, DataScope::Module, Result) then
            exit('');
```
to get some data that is stored and, if it returns a blank string, I am setting the values this way:

```
IsolatedStorage.Set(StrSubstNo('%1_%2', AppId, Name), KeyValue, DataScope::Module)
```
Basically using "AppId_Name" as the unique key name, since I am doing this in a library app that is used for all apps.

Although, I am the only user in the system, I am getting an error message "Another user has locked the table" or sometimes also an actual deadlock error. It is not completely clear how to replicate the issue, but it seems to be that, if I call the "Get" part, then the "Set" part, and then again the "Get" part at a later time, I am getting these errors.

**Comment by pzentner** [comment]
I understand that the error might come from the system keeping a lock on it, however, if I am the only user, it would either mean that some other processes in the background are using it and not using it correctly (possibly) or that the implementation of the functionality is wrong. When a write transaction implicitly is ended, it should commit everything to the database and actually end the transqction. A commit explicitly should not be needed. 

From a best practices perspective, it is also not very good to just put random commits in there, since this "set" could be called as part of a transaction, which then would be terminated prematurely, leaving the database in possibly an inconsistent state when an error occurs further down the line in the same transaction.

Therefore, I am reporting it here to review the implementation of IsolatedStorage as well as any other Microsoft apps that are part of the docker image and might be involved.

Based on my understanding, the way it appears is that IsolatedStorage accesses a system table in the background or we wouldnt see the database insert etc events being fired when that is called. Further, based on the behavior, it would seem that the write on the table is executed by the service tier in a different context and not in the context of the possibly started transaction. Those are obviously only assumptions, but if that's the case, it would have to be altered probably.

**Comment by PooyaKharamesh** [comment]
Hi 

Could you please provide us a full package that repros the issue?

/Pooya

**Comment by pzentner** [comment]
Unfortunately, I do not have the package anymore that I used to test the issue. But what I remember is that I only had the issue when I had two different transactions that were executed by subscribing to events and the both modified data in the Isolated Storage and they were executed quickly one after the other. I think it was the last event before the commit in codeunit 80 and then the first event after the commit. But I am not 100% sure anymore. But restructuring the code so that this can't happen (both modify quickly one after the other), I resolved the issue. So, it could be that there is some delay in completing the transaction or keeping the log on the IsolatedStorage.
